icefoganalytics/travel-authorization
# Standardize Vuex Store Pattern

Relates to:
- https://github.com/icefoganalytics/travel-authorization/issues/82

# Context

The `web/src/modules/travel-authorizations/store.js` store is getting pretty complex and unwieldy. I keep adding features, and am starting to lose track of what's doing what. Also the names are all pretty much random.

I'd done some investigation, and I think a good general pattern will be to break the "current" and generic stores apart. Also to separate stores by back-end model, rather than by specific module.

e.g.

- `/current/user`
- `/current/travel-authorization`
- `/users`
- `/travel-authorizations`
etc.

An example of a pretty clean current user store is:
<details>
<summary><code>web/src/store/current/current-user.js</code></summary>

```js
import usersApi from "@/api/users-api"

const state = {
  attributes: {},
  isLoading: false,
  isErrored: false,
  isInitialized: false,
}

const getters = {
  id: (state) => state.attributes.id,
  fullName: (state) => {
    const { firstName, lastName } = state.attributes
    return [firstName, lastName].filter(Boolean).join(" ")
  },
}

const actions = {
  async initialize({ commit }) {
    commit("SET_IS_LOADING", true)
    try {
      const { user } = await usersApi.me()
      commit("SET_IS_ERRORED", false)
      commit("SET_ATTRIBUTES", user)
      commit("SET_IS_INITIALIZED", true)
      return state.attributes
    } catch (error) {
      console.error("Failed to load current user:", error)
      commit("SET_IS_ERRORED", true)
      throw error
    } finally {
      commit("SET_IS_LOADING", false)
    }
  },
}

const mutations = {
  SET_ATTRIBUTES(state, value) {
    state.attributes = value
  },
  SET_IS_LOADING(state, value) {
    state.isLoading = value
  },
  SET_IS_ERRORED(state, value) {
    state.isErrored = value
  },
  SET_IS_INITIALIZED(state, value) {
    state.isInitialized = value
  },
}

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations,
}
```
</details>

Loaded in `web/src/store/current/index.js` as
```js
import user from "./current-user"

export default {
  namespaced: true,
  modules: {
    user,
  },
}
```

and in `web/src/store/index.js` as
```js
import Vue from "vue"
import Vuex from "vuex"

import current from "./current"

Vue.use(Vuex)

export default new Vuex.Store({
  modules: {
    current,
  }
})
```

resulting in method lookups like 
```js
export default {
  methods: {
    ...mapActions("current/user", { initializeCurrentUser: "initialize" }),
  }
}
```

# Implementation

Pattern for handling errors is logging, then re-throwing the error. In the future I'd like to switch to pushing the errors onto onto a notification queue and having a standard component show the errors.

# Testing Instructions

1. Boot the app via `dev up`
2. Log in to the app at http://localhost:8080.
3. Go to the My Travel Requests page via the top dropdown nav.
4. Check that you can view/create/update Travel Authorizations.
